src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2066,41 +2106,162 @@ TargetShardIntervalForModify(Query *query)
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	char partitionMethod = cacheEntry->partitionMethod;
+	bool fastShardPruningPossible = false;
+	uint32 rangeTableId = 1;
+	Var *partitionColumn = NULL;
+	Expr *partitionValueExpr = NULL;
+	Const *partitionValueConst = NULL;
+	List *shardIntervalList = NIL;
+	List *prunedShardList = NIL;
+	int prunedShardCount = 0;
+
+	Assert(query->commandType == CMD_INSERT);
+
+	/* reference tables can only have one shard */
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		int shardCount = 0;
+
+		shardIntervalList = LoadShardIntervalList(distributedTableId);
+		shardCount = list_length(shardIntervalList);
+
+		if (shardCount != 1)
+		{
+			ereport(ERROR, (errmsg("reference table cannot have %d shards", shardCount)));
+		}
+
+		return (ShardInterval *) linitial(shardIntervalList);
+	}
+
+	partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
+
+	partitionValueExpr = ExtractInsertPartitionValue(query, partitionColumn);
+	if (!IsA(partitionValueExpr, Const))
+	{
+		/* shard pruning not possible right now */
+		return NULL;
+	}
+
+	partitionValueConst = (Const *) partitionValueExpr;
+	if (partitionValueConst->constisnull)
+	{
+		ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+						errmsg("cannot perform an INSERT with NULL in the partition "
+							   "column")));
+	}
+
 	fastShardPruningPossible = FastShardPruningPossible(query->commandType,
 														partitionMethod);
 	if (fastShardPruningPossible)
 	{
-		uint32 rangeTableId = 1;
-		Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
-		Const *partitionValueConst = ExtractInsertPartitionValue(query, partitionColumn);
 		Datum partitionValue = partitionValueConst->constvalue;
 		ShardInterval *shardInterval = FastShardPruning(distributedTableId,
 														partitionValue);
-
 		if (shardInterval != NULL)
 		{
-			prunedShardList = lappend(prunedShardList, shardInterval);
+			prunedShardList = list_make1(shardInterval);
 		}
 	}
 	else
 	{
-		List *restrictClauseList = QueryRestrictList(query);
+		List *restrictClauseList = NIL;
 		Index tableId = 1;
-		List *shardIntervalList = LoadShardIntervalList(distributedTableId);
+		OpExpr *equalityExpr = MakeOpExpression(partitionColumn, BTEqualStrategyNumber);
+		Node *rightOp = get_rightop((Expr *) equalityExpr);
+		Const *rightConst = (Const *) rightOp;
+
+		Assert(IsA(rightOp, Const));
 
+		rightConst->constvalue = partitionValueConst->constvalue;
+		rightConst->constisnull = partitionValueConst->constisnull;
+		rightConst->constbyval = partitionValueConst->constbyval;
+
+		restrictClauseList = list_make1(equalityExpr);
+
+		shardIntervalList = LoadShardIntervalList(distributedTableId);
 		prunedShardList = PruneShardList(distributedTableId, tableId, restrictClauseList,
 										 shardIntervalList);
 	}
 
 	prunedShardCount = list_length(prunedShardList);
 	if (prunedShardCount != 1)
```

Comment:
I was thinking it's actually easier to understand without the mixing and matching of messages depending on the command type. I do feel like this function is a little strangely structured now that the insert logic is delegated to `FindShardForInsert`. I'm introducing a `FindShardForUpdateOrDelete` to make it more symmetric.